### PR TITLE
chose source/target branch to compare

### PR DESCRIPTION
**Pull request by yigger**
Both the source branch and the target branch allow users to choose freely.

### CODE REVIEW

**Comment by yigger** [comment]
I am thinking about add this feature to new command or modify diff branch in the original foundation? These two features look the same, and I think it is still modified on the original basis. But need to add small modifications that chose current branch as default when source branch quickpick appear . just like the below picture. How do you feel?

image

**Comment by huizhougit** [comment]
Many interactions to do branch diff might not be a good idea. It is not concise to user.  
I am thinking about using branchxxx..branchyyy for branch diff. So there are several scenarios:
  1. Input: branchxxx. 
      Output: diff between current branch and branchxxx.
  2. Input: branchxxx..branchyyy. 
      Output: diff between branchxxx and branchyyy.
  3. Input: Enter commit SHA -> shaxxx
      Output: diff between current branch and shaxxx.
  4. Input: Enter commit SHA -> shaxxx..shayyy
      Output: diff between shaxxx and shayyy

How do you think? @yigger

**Comment by yigger** [comment]
You means that do not change original interaction and support input value like 'sourcebranch..targetbranch' ?

If so, it may be annoying when user enter too many characters(branch name may too long).

So, I think that add new command to support the new feature and change original command's name to "View Branch Diff With Current Branch" If you don't want to make changes on the original basis.

Am I clear? @huizhougit

**Comment by huizhougit** [comment]
User doesn't need to have much typing. Just a few key characters is able to help popping up the selections and user could easily choose it.
The exception is SHA diff. Suppose user would copy/paste SHA anyway in that case.

image

**Comment by yigger** [comment]
Yeah. I got it.

**Comment by huizhougit** [comment]
@yigger Would you like to implement this feature as we discussed? I am planing to release a new version recently.

**Comment by yigger** [comment]
Yes, No.1 and No.2 use showQuickPick, but what's the showQuickPick items should be?
According to the plan you provided, The items should consist of the repo's branchs and comparison of each branch, right? But the problem is that the result of each branch comparison may be large, because the algorithm complexity is reached to 'n*(n-1)'

**Comment by yigger** [comment]
Such a rough calculation method will cause problems. This method may satisfy most small projects only. Middle or larger projects should have problems. For my current company's project, there are more than hundreds of branches, and there are multiple remo, the amount of data calculated is really too large, and exist too much useless data. I suggest to think about other solutions

**Comment by huizhougit** [comment]
Tried 10000 pickup items and it is very smooth. 
env: cpu i7 with 4 cores and 16g ram. 
I suppose your local branch * all is less than 10000.

**Comment by yigger** [comment]
Great, I tried to finish in these two days.

**Comment by yigger** [comment]
@huizhougit  
I am done. Review the code to see if there is any problem. 😊

File: src/commands.ts [hunk 1]
```diff
@@ -39,6 +39,24 @@ async function selectBranch(gitService: GitService, repo: GitRepo, allowEnterSha
...
+        return ref.type != GitRefType.Tag
+    });
+    const localRefs = refs.filter(ref => {
+        return ref.type != GitRefType.RemoteHead
+    })
+    let items: QuickPickItem[] = []
+    localRefs.forEach(source => {
+        refs.forEach(target => {
+            if (source.name != target.name && source.commit != target.commit) {
+                items.push({label: `${source.name || source.commit}..${target.name || target.commit}`})
```
Comment (huizhougit) [inline src/commands.ts]: Add ' '(space) before and after .. so it is easy to search two commits

File: src/commands.ts [hunk 2]
```diff
@@ -238,20 +256,32 @@ export class CommandCenter {
...
+            const branchs: QuickPickItem[] = await selectBranch(this._gitService, repo, true)
+            const branchWithCombination: QuickPickItem[] = await branchCombination(this._gitService, repo)
+            const items: QuickPickItem[] = [...branchs, ...branchWithCombination]
+            const placeHolder: string = 'Select a branch to compare...'
+            window.showQuickPick(items, { placeHolder: placeHolder }).then(async item => {
+                if (!item) {
+                    return;
+                }
+                const currentRef: string = await this._gitService.getCurrentBranch(repo);
+                let leftRef = await getRefFromQuickPickItem(item, `Input a ref(sha1) to compare with ${currentRef} or 'ref(sha1)..ref(sha2)'`);
```
Comment (huizhougit) [inline src/commands.ts]: Make the instruction more clear. Like
Input a ref(sha1) to compare with ${currentRef} or 'ref(sha1)..ref(sha2)' to compare with two commits

File: src/commands.ts [hunk 3]
```diff
@@ -238,20 +256,32 @@ export class CommandCenter {
...
+                let rightRef = currentRef;
+                if (!leftRef) return;
+
+                if (leftRef.indexOf('..') != -1) {
+                    const diffBranch = leftRef.split('..');
+                    leftRef = diffBranch[0];
```
Comment (huizhougit) [inline src/commands.ts]: don't forget trim() since we add spaces to ..

**Comment by yigger** [comment]
Merged master and format the code. @huizhougit

**Pull request merged**